vibi-dpu/src/graph/utils.rs

```diff
@@ -25,47 +27,72 @@ pub async fn call_llm_api(prompt: String) -> Option<String> {
+
+        let response_res = client
+            .post(url)
+            .header("api-key", format!("{}", token))
+            .header("Content-Type", "application/json")
+            .json(&json!({
+                "messages": [
+                    { "role": "user", "content": prompt }
+                ],
+                "max_tokens": 4000,
+            }))
+            .send()
+            .await;
+
+        if let Err(err) = response_res {
+            log::error!("[call_llm_api] Error in calling API: {:?}", err);
+            continue;
+        }
 
-    let resp_text = resp_text_res.unwrap();
-    // Parse the response text as JSON
-    let resp_json: Value = serde_json::from_str(&resp_text).expect("Failed to parse JSON");
-
-    // Extract and concatenate all `message.content` values from `choices`
-    let final_response: String = resp_json["choices"]
-        .as_array()
-        .unwrap_or(&vec![]) // Handle case where `choices` might be missing
-        .iter()
-        .filter_map(|choice| choice["message"]["content"].as_str()) // Extract content
-        .collect::<Vec<&str>>() // Collect into a Vec of strings
-        .join("\n"); // Join with newlines
-    log::info!("[call_llm_api] Final aggregated response: {:#?}", final_response);
-    Some(final_response)
+        let response = response_res.expect("Uncaught error in response_res");
+
+        // Ensure we can read the response stream
+        let resp_text_res = response.text().await;
+        if let Err(err) = resp_text_res {
+            log::error!("[call_llm_api] Error reading response text: {:?}", err);
+            continue;
+        }
+
+        let resp_text = resp_text_res.expect("Uncaught error in resp_text_res");
+        // Parse the response text as JSON
+        let resp_json_res = serde_json::from_str(&resp_text);
+        if let Err(e) = resp_json_res {
+            log::error!("[call_llm_api] Unable to parse response json {:?}", e);
+            continue;
+        }
+        let resp_json: Value = resp_json_res.expect("Failed to parse JSON");
+        // Extract and concatenate all `message.content` values from `choices`
+        let final_response: String = resp_json["choices"]
+            .as_array()
+            .unwrap_or(&vec![]) // Handle case where `choices` might be missing
+            .iter()
+            .filter_map(|choice| choice["message"]["content"].as_str()) // Extract content
+            .collect::<Vec<&str>>() // Collect into a Vec of strings
+            .join("\n"); // Join with newlines
+        log::info!("[call_llm_api] Final aggregated response: {:#?}", final_response);
+        if !final_response.is_empty() {
+            return Some(final_response);
+        }
+
+        log::warn!(
+            "[call_llm_api] Empty response, retrying in {} seconds (attempt {}/5)...",
+            wait_time,
+            attempt + 1
+        );
+        sleep(Duration::from_secs(wait_time)).await;
+        wait_time *= 2; // Double the wait time
```

Comment:
Oh, so in the case of rate-limiting, we get an empty response and not an error?

If that is tested and true, then this looks good to me.

Reply:
Yep, this is tested. When requests per minute is exhausted, this is the way to check